### PR TITLE
fix: catalog parsing for flake upgrades

### PR DESCRIPTION
**Pull request by limeytexan**
This commit partially reverts commit 7d16e9bb which broke the ability to detect available catalog entry upgrades and fixes the ability to detect available flake upgrades. It is being rolled out as a hotfix to flox-prerelease as this is currently breaking the functionality of `flox upgrade` in the majority of cases.

## Checks

- [x] All tests pass.
- [x] All commits in this Pull Request are [signed](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits) and Verified by Github.
- [x] I have accepted the flox [Contributor License Agreement](../blob/main/.github/CLA.md) by adding my Git/Github details in a row at the end of the [CONTRIBUTORS.csv](../blob/main/.github/CONTRIBUTORS.csv) file by way of this pull request or one done previously.

## Release Notes

* Fixed bug introduced with #74 which affected the ability to upgrade packages (both flakes and catalog packages) with `flox upgrade`.

### CODE REVIEW

File: flox-bash/lib/diff-catalogs.jq [hunk 1]
```diff
@@ -12,65 +12,88 @@ $c2[0] as $b |
...
-    to_entries | map(
-      (keys + [.key]) as $newkeys |
-      if (.value | isPackage) then keys else (
-          .value | packagePNames($newkeys)[]
-      ) end
+  to_entries | map(
+    (keys + [.key]) as $newkeys |
+    if ((.value | type) != "object") then (
+      # Uh oh ... something gone wrong, not a dict.
+      "ERROR: encountered object of type \(type)" | halt_error(1)
```
Comment (mkenigs) [inline flox-bash/lib/diff-catalogs.jq:74]:
```suggestion
      # We got to a leaf before finding a package, so there was never a valid catalog entry.
      "Internal error: invalid catalog data" | halt_error(1)
```

**Review by mkenigs — CHANGES_REQUESTED** [review]

File: flox-bash/lib/diff-catalogs.jq [hunk 2]
```diff
@@ -12,65 +12,88 @@ $c2[0] as $b |
...
+#           "eval": [ ... ],
+#           "version": 1
+#         }
+#       }
+#     }
+#   }
+# }
+#
+def isFlakePackage:
+  has("build") and has("element") and has("eval") and has("version");
```
[the remark this answers is not in the document]
Comment (limeytexan) [reply flox-bash/lib/diff-catalogs.jq:66]: Thanks. Amending accordingly.

```suggestion
  (has("type") and .["type"] == "catalogRender") or (has("element") and has("eval") and has("version"));
```

File: flox-bash/lib/diff-catalogs.jq [hunk 3]
```diff
@@ -12,65 +12,88 @@ $c2[0] as $b |
...
+#           "element": [ ... ],
+#           "eval": [ ... ],
+#           "version": 1
+#         }
+#       }
+#     }
+#   }
+# }
+#
+def isFlakePackage:
```
Comment (limeytexan) [inline flox-bash/lib/diff-catalogs.jq:65]:
```suggestion
def isPackage:
```

File: flox-bash/lib/diff-catalogs.jq [hunk 4]
```diff
@@ -12,65 +12,88 @@ $c2[0] as $b |
...
+#   }
+# }
+def isCatalogPackage:
+  (
+    has("build") and has("element") and has("eval") and has("version") and
+    has("cache") and has("publish_element") and has("source")
+  ) or (
+    # TODO: remove, "type" is deprecated?
+    has("type") and .["type"] == "catalogRender"
+  );
```
Comment (limeytexan) [inline flox-bash/lib/diff-catalogs.jq:47]:
```suggestion
  isPackage and has("publish_element");
```

File: flox-bash/lib/diff-catalogs.jq [hunk 5]
```diff
@@ -12,65 +12,88 @@ $c2[0] as $b |
...
-    false
-  ) end;
+def isCatalogPackage:
+  (
+    has("build") and has("element") and has("eval") and has("version") and
+    has("cache") and has("publish_element") and has("source")
+  ) or (
+    # TODO: remove, "type" is deprecated?
+    has("type") and .["type"] == "catalogRender"
+  );
```
[the remark this answers is not in the document]
Comment (limeytexan) [reply flox-bash/lib/diff-catalogs.jq]: Done.

**Comment by limeytexan** [comment]
@mkenigs thanks for the review - I've updated accordingly can you give it one final review? Thanks.

**Review by mkenigs — APPROVED** [review]

**Comment by ysndr** [comment]
@limeytexan is this in for the release today?

**Comment by limeytexan** [comment]
Thanks for flagging - yes let's go ahead and roll this one in.

**Pull request merged**